File: sources/iOS/ios-communications/Tests/iOSCommunicationsTests/PmdOfflineTriggerTest.swift
//  Copyright © 2023 Polar. All rights reserved.

import XCTest
@testable import iOSCommunications

final class PmdOfflineTriggerTest: XCTestCase {
    
    func testTriggersAreDisabled() throws {
        //Arrange
        let pmdGetTriggerStatusResponse = Data([
            // index 0 : trigger mode
            0x00,
            // index 1.. : trigger status, measurement type, settings length (optional), settings (optional)
            0x00, 0x02,
            0x00, 0x05,
            0x00, 0x06,
            0x00, 0x01,
            0x00, 0x03
        ])
        
        //Act
        let pmdOfflineTrigger = try PmdOfflineTrigger.fromResponse(data: pmdGetTriggerStatusResponse)
        
        //Assert
        XCTAssertEqual(PmdOfflineRecTriggerMode.disabled, pmdOfflineTrigger.triggerMode)
        XCTAssertTrue(pmdOfflineTrigger.triggers.contains(where: { $0.key == PmdMeasurementType.ppi}))
        XCTAssertTrue(pmdOfflineTrigger.triggers.contains(where: { $0.key == PmdMeasurementType.acc}))
        XCTAssertTrue(pmdOfflineTrigger.triggers.contains(where: { $0.key == PmdMeasurementType.ppg}))
        XCTAssertTrue(pmdOfflineTrigger.triggers.contains(where: { $0.key == PmdMeasurementType.gyro}))
        XCTAssertTrue(pmdOfflineTrigger.triggers.contains(where: { $0.key == PmdMeasurementType.mgn}))
        XCTAssertTrue(pmdOfflineTrigger.triggers.count == 5)
        for trigger in pmdOfflineTrigger.triggers {
            XCTAssertTrue(trigger.value.status == PmdOfflineRecTriggerStatus.disabled)
            XCTAssertTrue(trigger.value.setting == nil)
        }
    }
    
    func testTriggerPpiEnabledAtSystemStart() throws {
        //Arrange
        let pmdGetTriggerStatusResponse = Data([
            // index 0 : trigger mode
            0x01,
            // index 1.. : trigger status, measurement type, settings length (optional), settings (optional)
            0x00, 0x02,
            0x00, 0x05,
            0x00, 0x06,
            0x00, 0x01,
            0x01, 0x03, 0x00
        ])
        
        //Act
        let pmdOfflineTrigger = try PmdOfflineTrigger.fromResponse(data:pmdGetTriggerStatusResponse)
        
        //Assert
        XCTAssertEqual(PmdOfflineRecTriggerMode.systemStart, pmdOfflineTrigger.triggerMode)
        XCTAssertTrue(pmdOfflineTrigger.triggers.contains(where: { $0.key == PmdMeasurementType.ppi}))
        XCTAssertTrue(pmdOfflineTrigger.triggers.contains(where: { $0.key == PmdMeasurementType.acc}))
        XCTAssertTrue(pmdOfflineTrigger.triggers.contains(where: { $0.key == PmdMeasurementType.ppg}))
        XCTAssertTrue(pmdOfflineTrigger.triggers.contains(where: { $0.key == PmdMeasurementType.gyro}))
        XCTAssertTrue(pmdOfflineTrigger.triggers.contains(where: { $0.key == PmdMeasurementType.mgn}))
        XCTAssertTrue(pmdOfflineTrigger.triggers.count == 5)
        for trigger in pmdOfflineTrigger.triggers {
            if (trigger.key == PmdMeasurementType.ppi) {
                XCTAssertTrue(trigger.value.status == PmdOfflineRecTriggerStatus.enabled)
                XCTAssertTrue(trigger.value.setting == nil)
            } else {
                XCTAssertTrue(trigger.value.status == PmdOfflineRecTriggerStatus.disabled)
                XCTAssertTrue(trigger.value.setting == nil)
            }
        }
    }
    
    func testTriggerPpiAccGyroMagPpgEnabledAtExerciseStart() throws {
        //Arrange
        let pmdGetTriggerStatusResponse = Data([
            // index 0 : trigger mode
            0x02,
            // index 1.. : trigger status, measurement type, settings length (optional), settings (optional)
            0x01, 0x02, 0x0f, 0x00, 0x01, 0x34, 0x00, 0x01, 0x01, 0x10, 0x00, 0x02, 0x01, 0x08, 0x00, 0x04, 0x01, 0x03,
            0x01, 0x05, 0x0f, 0x00, 0x01, 0x34, 0x00, 0x01, 0x01, 0x10, 0x00, 0x02, 0x01, 0xd0, 0x07, 0x04, 0x01, 0x03,
            0x01, 0x06, 0x0f, 0x00, 0x01, 0x64, 0x00, 0x01, 0x01, 0x10, 0x00, 0x02, 0x01, 0x32, 0x00, 0x04, 0x01, 0x03,
            0x01, 0x01, 0x0f, 0x00, 0x01, 0x87, 0x00, 0x01, 0x01, 0x16, 0x00, 0x02, 0x01, 0x00, 0x00, 0x04, 0x01, 0x04,
            0x01, 0x03, 0x00
        ])
        
        //Act
        let pmdOfflineTrigger = try PmdOfflineTrigger.fromResponse(data:pmdGetTriggerStatusResponse)
        
        //Assert
        XCTAssertEqual(PmdOfflineRecTriggerMode.exerciseStart, pmdOfflineTrigger.triggerMode)
        XCTAssertTrue(pmdOfflineTrigger.triggers.contains(where: { $0.key == PmdMeasurementType.ppi}))
        XCTAssertTrue(pmdOfflineTrigger.triggers.contains(where: { $0.key == PmdMeasurementType.acc}))
        XCTAssertTrue(pmdOfflineTrigger.triggers.contains(where: { $0.key == PmdMeasurementType.ppg}))
        XCTAssertTrue(pmdOfflineTrigger.triggers.contains(where: { $0.key == PmdMeasurementType.gyro}))
        XCTAssertTrue(pmdOfflineTrigger.triggers.contains(where: { $0.key == PmdMeasurementType.mgn}))
        
        XCTAssertTrue(pmdOfflineTrigger.triggers.count == 5)
        for trigger in pmdOfflineTrigger.triggers {
            if (trigger.key == PmdMeasurementType.ppi) {
                XCTAssertTrue(trigger.value.status == PmdOfflineRecTriggerStatus.enabled)
                XCTAssertTrue(trigger.value.setting == nil)
            } else {
                XCTAssertTrue(trigger.value.status == PmdOfflineRecTriggerStatus.enabled)
                XCTAssertTrue(trigger.value.setting != nil)
            }
        }
    }
    
    func testTriggerPmdResponseHasWrongStatus() throws {
        //Arrange
        let pmdGetTriggerStatusResponse = Data([
            // index 0 : trigger mode
            0x01,
            // index 1.. : trigger status, measurement type, settings length (optional), settings (optional)
            0x03, 0x01
        ])
        //Act & Assert
        XCTAssertThrowsError(
            try PmdOfflineTrigger.fromResponse(data: pmdGetTriggerStatusResponse))
        { error in
            guard case BleGattException.gattDataError = error else {
                return XCTFail()
            }
        }
    }
}

File: sources/Android/android-communications/library/src/test/java/com/polar/androidcommunications/api/ble/model/gatt/client/pmd/PmdOfflineTriggerTest.kt
package com.polar.androidcommunications.api.ble.model.gatt.client.pmd

import com.polar.androidcommunications.api.ble.exceptions.BleControlPointResponseError
import com.polar.androidcommunications.testrules.BleLoggerTestRule
import org.junit.Assert.*
import org.junit.Rule
import org.junit.Test

internal class PmdOfflineTriggerTest {
    @Rule
    @JvmField
    val bleLoggerTestRule = BleLoggerTestRule()

    @Test
    fun `test triggers are disabled`() {
        //Arrange
        val pmdGetTriggerStatusResponse = byteArrayOf(
            // index 0 : trigger mode
            0x00.toByte(),
            // index 1.. : trigger status, measurement type, settings length (optional), settings (optional)
            0x00.toByte(), 0x02.toByte(),
            0x00.toByte(), 0x05.toByte(),
            0x00.toByte(), 0x06.toByte(),
            0x00.toByte(), 0x01.toByte(),
            0x00.toByte(), 0x03.toByte()
        )

        //Act
        val pmdOfflineTrigger = PmdOfflineTrigger.parseFromResponse(pmdGetTriggerStatusResponse)

        //Assert
        assertEquals(PmdOfflineRecTriggerMode.TRIGGER_DISABLE, pmdOfflineTrigger.triggerMode)
        assertTrue(pmdOfflineTrigger.triggers.keys.contains(PmdMeasurementType.PPI))
        assertTrue(pmdOfflineTrigger.triggers.keys.contains(PmdMeasurementType.ACC))
        assertTrue(pmdOfflineTrigger.triggers.keys.contains(PmdMeasurementType.PPG))
        assertTrue(pmdOfflineTrigger.triggers.keys.contains(PmdMeasurementType.GYRO))
        assertTrue(pmdOfflineTrigger.triggers.keys.contains(PmdMeasurementType.MAGNETOMETER))
        assertTrue(pmdOfflineTrigger.triggers.keys.size == 5)
        pmdOfflineTrigger.triggers.values.map {
            assertTrue(it.first == PmdOfflineRecTriggerStatus.TRIGGER_DISABLED)
            assertTrue(it.second == null)
        }
    }

    @Test
    fun `test trigger PPI enabled at system start`() {
        //Arrange
        val pmdGetTriggerStatusResponse = byteArrayOf(
            // index 0 : trigger mode
            0x01.toByte(),
            // index 1.. : trigger status, measurement type, settings length (optional), settings (optional)
            0x00.toByte(), 0x02.toByte(),
            0x00.toByte(), 0x05.toByte(),
            0x00.toByte(), 0x06.toByte(),
            0x00.toByte(), 0x01.toByte(),
            0x01.toByte(), 0x03.toByte(), 0x00.toByte()
        )

        //Act
        val pmdOfflineTrigger = PmdOfflineTrigger.parseFromResponse(pmdGetTriggerStatusResponse)

        //Assert
        assertEquals(PmdOfflineRecTriggerMode.TRIGGER_SYSTEM_START, pmdOfflineTrigger.triggerMode)
        assertTrue(pmdOfflineTrigger.triggers.keys.contains(PmdMeasurementType.PPI))
        assertTrue(pmdOfflineTrigger.triggers.keys.contains(PmdMeasurementType.ACC))
        assertTrue(pmdOfflineTrigger.triggers.keys.contains(PmdMeasurementType.PPG))
        assertTrue(pmdOfflineTrigger.triggers.keys.contains(PmdMeasurementType.GYRO))
        assertTrue(pmdOfflineTrigger.triggers.keys.contains(PmdMeasurementType.MAGNETOMETER))
        assertTrue(pmdOfflineTrigger.triggers.keys.size == 5)
        pmdOfflineTrigger.triggers.map {
            if (it.key == PmdMeasurementType.PPI) {
                assertTrue(it.value.first == PmdOfflineRecTriggerStatus.TRIGGER_ENABLED)
                assertTrue(it.value.second == null)
            } else {
                assertTrue(it.value.first == PmdOfflineRecTriggerStatus.TRIGGER_DISABLED)
                assertTrue(it.value.second == null)
            }
        }
    }

    @Test
    fun `test trigger PPI, ACC, GYRO, MAG, PPG enabled at exercise start`() {
        //Arrange
        val pmdGetTriggerStatusResponse = byteArrayOf(
            // index 0 : trigger mode
            0x02.toByte(),
            // index 1.. : trigger status, measurement type, settings length (optional), settings (optional)
            0x01.toByte(), 0x02.toByte(), 0x0f.toByte(), 0x00.toByte(), 0x01.toByte(), 0x34.toByte(), 0x00.toByte(), 0x01.toByte(), 0x01.toByte(), 0x10.toByte(), 0x00.toByte(), 0x02.toByte(), 0x01.toByte(), 0x08.toByte(), 0x00.toByte(), 0x04.toByte(), 0x01.toByte(), 0x03.toByte(),
            0x01.toByte(), 0x05.toByte(), 0x0f.toByte(), 0x00.toByte(), 0x01.toByte(), 0x34.toByte(), 0x00.toByte(), 0x01.toByte(), 0x01.toByte(), 0x10.toByte(), 0x00.toByte(), 0x02.toByte(), 0x01.toByte(), 0xd0.toByte(), 0x07.toByte(), 0x04.toByte(), 0x01.toByte(), 0x03.toByte(),
            0x01.toByte(), 0x06.toByte(), 0x0f.toByte(), 0x00.toByte(), 0x01.toByte(), 0x64.toByte(), 0x00.toByte(), 0x01.toByte(), 0x01.toByte(), 0x10.toByte(), 0x00.toByte(), 0x02.toByte(), 0x01.toByte(), 0x32.toByte(), 0x00.toByte(), 0x04.toByte(), 0x01.toByte(), 0x03.toByte(),
            0x01.toByte(), 0x01.toByte(), 0x0f.toByte(), 0x00.toByte(), 0x01.toByte(), 0x87.toByte(), 0x00.toByte(), 0x01.toByte(), 0x01.toByte(), 0x16.toByte(), 0x00.toByte(), 0x02.toByte(), 0x01.toByte(), 0x00.toByte(), 0x00.toByte(), 0x04.toByte(), 0x01.toByte(), 0x04.toByte(),
            0x01.toByte(), 0x03.toByte(), 0x00.toByte()
        )
        //Act
        val pmdOfflineTrigger = PmdOfflineTrigger.parseFromResponse(pmdGetTriggerStatusResponse)

        //Assert
        assertEquals(PmdOfflineRecTriggerMode.TRIGGER_EXERCISE_START, pmdOfflineTrigger.triggerMode)
        assertTrue(pmdOfflineTrigger.triggers.keys.contains(PmdMeasurementType.PPI))
        assertTrue(pmdOfflineTrigger.triggers.keys.contains(PmdMeasurementType.ACC))
        assertTrue(pmdOfflineTrigger.triggers.keys.contains(PmdMeasurementType.PPG))
        assertTrue(pmdOfflineTrigger.triggers.keys.contains(PmdMeasurementType.GYRO))
        assertTrue(pmdOfflineTrigger.triggers.keys.contains(PmdMeasurementType.MAGNETOMETER))
        assertTrue(pmdOfflineTrigger.triggers.keys.size == 5)
        pmdOfflineTrigger.triggers.map {
            if (it.key == PmdMeasurementType.PPI) {
                assertTrue(it.value.first == PmdOfflineRecTriggerStatus.TRIGGER_ENABLED)
                assertTrue(it.value.second == null)
            } else {
                assertTrue(it.value.first == PmdOfflineRecTriggerStatus.TRIGGER_ENABLED)
                assertTrue(it.value.second != null)
            }
        }
    }

    @Test
    fun `test trigger PMD response has wrong status`() {
        //Arrange
        val pmdGetTriggerStatusResponse = byteArrayOf(
            // index 0 : trigger mode
            0x01.toByte(),
            // index 1.. : trigger status, measurement type, settings length (optional), settings (optional)
            0x03.toByte(), 0x01.toByte()
        )
        //Act & Assert
        assertThrows(BleControlPointResponseError::class.java) {
            PmdOfflineTrigger.parseFromResponse(pmdGetTriggerStatusResponse)
        }
    }
}
